lib/ohai/plugins/ec2.rb

```diff
@@ -48,25 +65,26 @@ def has_ec2_dmi?
     begin
       # detect a version of '4.2.amazon'
       if dmi[:bios][:all_records][0][:Version] =~ /amazon/
-        Ohai::Log.debug("has_ec2_dmi? == true")
+        Ohai::Log.debug("ec2 plugin: has_ec2_dmi? == true")
         true
       end
     rescue NoMethodError
-      Ohai::Log.debug("has_ec2_dmi? == false")
+      Ohai::Log.debug("ec2 plugin: has_ec2_dmi? == false")
       false
     end
   end
 
-
   def looks_like_ec2?
-    # Try non-blocking connect so we don't "block" if
+    return true if hint?('ec2')
+
+    # if has ec2 mac try non-blocking connect so we don't "block" if
     # the Xen environment is *not* EC2
-    hint?('ec2') || ( has_ec2_dmi? || has_ec2_mac?) && can_metadata_connect?(Ohai::Mixin::Ec2Metadata::EC2_METADATA_ADDR,80)
+    return true if (has_ec2metadata_bin? || has_ec2_dmi?) || (has_xen_mac? && can_metadata_connect?(Ohai::Mixin::Ec2Metadata::EC2_METADATA_ADDR,80))
```

Comment:
The original logic and the new logic here look different to me (but I'm having trouble parsing it). It looks like the original logic would check `can_metadata_connect?` unless a hint was provided. The new logic only checks `can_metadata_connect?` if `has_xen_mac?` is true and both `has_ec2metadata_bin?` and `has_ec2_dmi?` are false. The variation I'm seeing here is that the new logic doesn't check `can_metadata_connect?` when `has_ec2_dmi?` is true.